pkg/providers/gcp/provider_postgres.go

```diff
@@ -125,12 +178,64 @@ func (p *PostgresProvider) ReconcilePostgres(ctx context.Context, pg *v1alpha1.P
 	}
 	logger.Infof("created network service connection %s", service.Service)
 
-	// TODO implement me
-	return p.createCloudSQLInstance(ctx, pg)
+	return p.reconcileCloudSQLInstance(ctx, pg, sqlClient, cloudSQLCreateConfig, strategyConfig, maintenanceWindow)
 }
 
-func (p *PostgresProvider) createCloudSQLInstance(ctx context.Context, pg *v1alpha1.Postgres) (*providers.PostgresInstance, croType.StatusMessage, error) {
-	return nil, croType.StatusEmpty, nil
+func (p *PostgresProvider) reconcileCloudSQLInstance(ctx context.Context, pg *v1alpha1.Postgres, sqladminService gcpiface.SQLAdminService, cloudSQLCreateConfig *sqladmin.DatabaseInstance, strategyConfig *StrategyConfig, maintenanceWindow bool) (*providers.PostgresInstance, croType.StatusMessage, error) {
+	logger := p.Logger.WithField("action", "reconcileCloudSQLInstance")
+	logger.Infof("reconciling cloudSQL instance")
+
+	instances, err := getCloudSQLInstances(sqladminService, strategyConfig.ProjectID)
+	if err != nil {
+		msg := "cannot retrieve sql instances from gcp"
+		return nil, croType.StatusMessage(msg), errorUtil.Wrap(err, msg)
+	}
+	logrus.Info("listed sql instances from gcp")
+
+	credSec := &v1.Secret{}
+	if err := p.Client.Get(ctx, types.NamespacedName{Name: pg.Name + defaultCredSecSuffix, Namespace: pg.Namespace}, credSec); err != nil {
+		msg := "failed to retrieve cloudSQL credential secret"
+		return nil, croType.StatusMessage(msg), errorUtil.Wrap(err, msg)
+	}
+
+	postgresPass := string(credSec.Data[defaultPostgresPasswordKey])
+	if postgresPass == "" {
+		msg := "unable to retrieve postgres password"
+		return nil, croType.StatusMessage(msg), errorUtil.Errorf(msg)
+	}
```

Comment:
Is there a need to check for the cloudsql secret right after creating it? I think we can safely remove that code block as the previous call to `CreateOrUpdate` would ensure that it exists.

Reply:
Removed.